src/sql/statement.rs

```diff
@@ -683,8 +672,74 @@ fn handle_create_sink(scx: &StatementContext, stmt: Statement) -> Result<Plan, f
             topic_prefix,
             topic_suffix,
         }),
+    })
+}
+
+fn handle_create_avro_ocf_sink(
+    scx: &StatementContext,
+    create_sql: String,
+    from: ObjectName,
+    path: String,
+    format: Option<Format>,
+) -> Result<Sink, failure::Error> {
+    if format.is_some() {
+        bail!("avro ocf sinks cannot specify a format");
+    }
+
+    let from = scx.resolve_name(from)?;
+    let catalog_entry = scx.catalog.get(&from)?;
+
+    let path = PathBuf::from(path);
+
+    if path.is_dir() {
+        bail!("avro ocf sink cannot write to a directory");
+    }
+
+    let creation_time = scx
+        .catalog
+        .creation_time()
+        .duration_since(UNIX_EPOCH)?
+        .as_secs();
+    let nonce = scx.catalog.nonce();
+
+    let file_name_suffix = format!("{}-{}", creation_time, nonce);
+
+    Ok(Sink {
+        create_sql,
+        from: catalog_entry.id(),
+        connector_builder: SinkConnectorBuilder::AvroOcf(AvroOcfSinkConnectorBuilder {
+            path,
+            file_name_suffix,
+        }),
+    })
+}
+
+fn handle_create_sink(scx: &StatementContext, stmt: Statement) -> Result<Plan, failure::Error> {
+    let create_sql = normalize::create_statement(scx, stmt.clone())?;
+    let (name, from, connector, format, if_not_exists) = match stmt {
+        Statement::CreateSink {
+            name,
+            from,
+            connector,
+            format,
+            if_not_exists,
+        } => (name, from, connector, format, if_not_exists),
+        _ => unreachable!(),
+    };
+
+    let sink = match connector {
+        Connector::File { .. } => bail!("file sinks are not yet supported"),
+        Connector::Kafka { broker, topic } => {
+            handle_create_kafka_sink(scx, create_sql, from, broker, topic, format)?
+        }
+        Connector::Kinesis { .. } => bail!("Kinesis sinks are not yet supported"),
+        Connector::AvroOcf { path } => {
+            handle_create_avro_ocf_sink(scx, create_sql, from, path, format)?
+        }
     };
 
+    let name = scx.allocate_name(normalize::object_name(name)?);
```

Comment:
Looks good, but you can eliminate a lot of duplicate code if you make the per-type functions just responsible for creating the builder, rather than the whole sink:

```rust
fn kafka_sink_builder(
    format: Option<Format>,
    topic_suffix: String,
    from_desc: RelationDesc,
    mut broker: String,
    topic_prefix: String,
) -> Result<SinkConnectorBuilder, failure::Error> {
    let schema_registry_url = match format {
        Some(Format::Avro(AvroSchema::CsrUrl { url, seed })) => {
            if seed.is_some() {
                bail!("SEED option does not make sense with sinks");
            }
            url.parse()?
        }
        _ => bail!("only confluent schema registry avro sinks are supported"),
    };

    if !broker.contains(':') {
        broker += ":9092";
    }
    let broker_url = broker.parse()?;

    let encoder = Encoder::new(from_desc);
    let value_schema = encoder.writer_schema().canonical_form();

    Ok(SinkConnectorBuilder::Kafka(KafkaSinkConnectorBuilder {
        broker_url,
        schema_registry_url,
        value_schema,
        topic_prefix,
        topic_suffix,
    }))
}

fn avro_ocf_sink_builder(
    format: Option<Format>,
    file_name_suffix: String,
    path: String,
) -> Result<SinkConnectorBuilder, failure::Error> {
    if format.is_some() {
        bail!("avro ocf sinks cannot specify a format");
    }

    let path = PathBuf::from(path);
    if path.is_dir() {
        bail!("avro ocf sink cannot write to a directory");
    }

    Ok(SinkConnectorBuilder::AvroOcf(AvroOcfSinkConnectorBuilder {
        path,
        file_name_suffix,
    }))
}

fn handle_create_sink(scx: &StatementContext, stmt: Statement) -> Result<Plan, failure::Error> {
    let create_sql = normalize::create_statement(scx, stmt.clone())?;
    let (name, from, connector, format, if_not_exists) = match stmt {
        Statement::CreateSink {
            name,
            from,
            connector,
            format,
            if_not_exists,
        } => (name, from, connector, format, if_not_exists),
        _ => unreachable!(),
    };

    let name = scx.allocate_name(normalize::object_name(name)?);
    let from = scx.catalog.get(&scx.resolve_name(from)?)?;
    let suffix = format!(
        "{}-{}",
        scx.catalog
            .creation_time()
            .duration_since(UNIX_EPOCH)?
            .as_secs(),
        scx.catalog.nonce()
    );

    let connector_builder = match connector {
        Connector::File { .. } => bail!("file sinks are not yet supported"),
        Connector::Kafka { broker, topic } => {
            kafka_sink_builder(format, suffix, from.desc()?.clone(), broker, topic)?
        }
        Connector::Kinesis { .. } => bail!("Kinesis sinks are not yet supported"),
        Connector::AvroOcf { path } => avro_ocf_sink_builder(format, suffix, path)?,
    };

    Ok(Plan::CreateSink {
        name,
        sink: Sink {
            create_sql,
            from: from.id(),
            connector_builder,
        },
        if_not_exists,
    })
}
```

(I've learned to try these things before suggesting them because Rust's borrowing rules make it very easy to suggest something that doesn't work—apologies for pasting a whole lot of code at you.)

Reply:
no worries this was a great suggestion and I appreciate that you tried it out :)